src/v/kafka/server/handlers/produce.cc

```diff
@@ -727,6 +721,58 @@ produce_handler::handle(request_context ctx, ss::smp_service_group ssg) {
         return process_result_stages::single_stage(ctx.respond(
           request.make_error_response(error_code::invalid_required_acks)));
     }
+
+    // Must now validate if we are authorized, we will remove items from the
+    // request that are not authorized and create response entries for those.
+    // Once authz is checked, then attempt audit
+    auto unauthorized_it = std::partition(
+      request.data.topics.begin(),
+      request.data.topics.end(),
+      [&ctx](const topic_produce_data& t) {
+          return ctx.authorized(security::acl_operation::write, t.name);
+      });
+
+    const auto unauthorized_contains_topic = [&unauthorized_it,
+                                              &request](const model::topic& t) {
+        return std::any_of(
+          unauthorized_it,
+          request.data.topics.end(),
+          [t](const topic_produce_data& tp) { return tp.name == t; });
+    };
+
+    // We do not want to audit if the request contains the audit topic
+    // however we _definitely_ want to audit if it was contained in the
+    // unauthorized list
+    if (
+      !(ctx.request_contains_audit_topic()
```

Comment:
nit: Although effectively the same would it be better to check if the client is an ephemeral user when deciding to not audit? That way we could remove this request_contains_audit_logic check altogether. This check is in a future PR i have lined up though. Maybe we do it later

Reply:
True, and I think in a round-about-way this is doing that by checking the `unauthorized_contains_topic` method.

But I think we're going in a different direction anyways when it comes to how to protect this topic.  Maybe we handle it in a follow up PR?